scvi/inference/experimental_inference.py

```diff
@@ -0,0 +1,89 @@
+import types
+from math import sqrt, pi
+
+import torch
+import torch.nn.functional as F
+from torch.distributions import Normal, Uniform
+
+from scvi.models.classifier import Classifier
+
+
+def mmd_fourier(x1, x2, bandwidth=2., dim_r=500):
+    d = x1.size(1)
+    rw_n = sqrt(2. / bandwidth) * Normal(0., 1. / sqrt(d)).sample((dim_r, d)).type(x1.type())
+    rb_u = 2 * pi * Uniform(0., 1.).sample((dim_r,)).type(x1.type())
+    rf0 = sqrt(2. / dim_r) * torch.cos(F.linear(x1, rw_n, rb_u))
+    rf1 = sqrt(2. / dim_r) * torch.cos(F.linear(x2, rw_n, rb_u))
+    result = (torch.pow(rf0.mean(dim=0) - rf1.mean(dim=0), 2)).sum()
+    return torch.sqrt(result)
+
+
+def mmd_objective(z, batch_index, n_batch):
+    mmd_method = mmd_fourier
+
+    z_dim = z.size(1)
+    batch_index = batch_index.view(-1)
+
+    # STEP 1: construct lists of samples in their proper batches
+    z_part = [z[batch_index == b_i] for b_i in range(n_batch)]
+
+    # STEP 2: add noise to all of them and get the mmd
+    mmd = 0
+    for j, z_j in enumerate(z_part):
+        z0_ = z_j
+        aux_z0 = Normal(0., 1.).sample((1, z_dim)).type(z0_.type())
+        z0 = torch.cat((z0_, aux_z0), dim=0)
+        if len(z_part) == 2:
+            z1_ = z_part[j + 1]
+            aux_z1 = Normal(0., 1.).sample((1, z_dim)).type(z1_.type())
+            z1 = torch.cat((z1_, aux_z1), dim=0)
+            return mmd_method(z0, z1)
+        z1 = z
+        mmd += mmd_method(z0, z1)
+    return mmd
+
+
+def mmd_loss(self, tensors, *next_tensors):
+    if self.epoch > self.warm_up:  # Leave a warm-up
+        sample_batch, _, _, batch_index, label = tensors
+        qm_z, _, _ = self.model.z_encoder(torch.log(1 + sample_batch), label)  # label only used in VAEC
+        loss = mmd_objective(qm_z, batch_index, self.gene_dataset.n_batches)
+    return type(self).loss(self, tensors, *next_tensors) + loss
+
+
+def mmd_wrapper(infer, warm_up=100, scale=50):
+    infer.warm_up = warm_up
+    infer.scale = scale
+    infer.loss = types.MethodType(adversarial_loss, infer)
+    infer.train = types.MethodType(adversarial_train, infer)
```

Comment:
This isn't a particularly idiomatic way to override the loss and train methods. It makes sense as far as getting something working, and it's pretty concise, but I worry it'll be hard to maintain.

Reply:
Would it work to make a `VariationalInferenceMMD` class that inherits from `VariationalInference` instead, which overrides `train` and `loss`?

Reply:
I wonder if it was a mistake to make a hierarchy of `Inference` classes, rather than using doing something more like functional programming; i.e., calling functions with other functions are arguments. So, for example, a `train` function might be called with `loss` function as an argument. The functional programming approach might have made it easier to implement everything we want, without much boilerplate and without ever duplicating code.

Just something to think about.  Probably better not to go down that rabbit hole yet. Please merge this PR whenever you're ready.

Reply:
I see, ok I will keep that in mind.